internal/pipeline/step.go

```diff
@@ -85,3 +93,113 @@ func (si *StartInformers) Cancel() error {
 	si.Status("cancel step")
 	return nil
 }
+
+type StartWatcher struct {
+	pipeline.StepContext
+	stopChan    chan struct{}
+	dynamicKube dynamic.Interface
+	log         logger.Handler
+	broker      broker.Handler
+	config      internalconfig.PipelineConfig
+}
+
+func newStartWatcherStage(dynamicKube dynamic.Interface, config internalconfig.PipelineConfig, stopChan chan struct{}, log logger.Handler, broker broker.Handler) *StartWatcher {
+	return &StartWatcher{
+		stopChan:    stopChan,
+		dynamicKube: dynamicKube,
+		config:      config,
+		log:         log,
+		broker:      broker,
+	}
+}
+
+func (w *StartWatcher) Exec(request *pipeline.Request) *pipeline.Result {
+	b := true
+	opts := metav1.ListOptions{
+		Watch:                true,
+		SendInitialEvents:    &b,
+		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+	}
+	// attempts to begin watching the namespaces
+	// returns a `watch.Interface`, or an error
+
+	gvr, _ := schema.ParseResourceArg(w.config.Name)
+	watcher, err := w.dynamicKube.Resource(*gvr).Watch(context.TODO(), opts)
+	if err != nil {
+		return &pipeline.Result{
+			Error: err,
+			Data:  nil,
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	// Launch the goroutine and pass the channel as an argument
+	wg.Add(1)
+	go w.backgroundWatchProcessor(watcher.ResultChan(), w.stopChan, &wg)
+	data := make(map[string]cache.Store)
+
+	return &pipeline.Result{
+		Error: nil,
+		Data:  data,
+	}
+}
+
+func (w *StartWatcher) backgroundWatchProcessor(result <-chan watch.Event, stopCh chan struct{}, wg *sync.WaitGroup) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			for event := range result {
+				obj := event.Object
+				switch event.Type {
+				// when an event is added...
+				case watch.Added:
+					err := w.publishItem(obj.(*unstructured.Unstructured), broker.Add, w.config)
+					if err != nil {
+						w.log.Error(err)
+					}
+					w.log.Info("Received ADD event for: ", obj.(*unstructured.Unstructured).GetName(), "/", obj.(*unstructured.Unstructured).GetNamespace(), " of kind: ", obj.(*unstructured.Unstructured).GroupVersionKind().Kind)
+				// when an event is modified...
+				case watch.Modified:
+					err := w.publishItem(obj.(*unstructured.Unstructured), broker.Update, w.config)
```

Comment:
For the Update functionality, using informers, one could get the older object and the new object and hence we could compare resource versions to determine whether indeed this object is a valid update or a repeated action. This is not available when using watchers